runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/FileWatcherTask.java

```diff
@@ -0,0 +1,169 @@
+    private NamespaceConfig rootNamespace;
+    private byte[] configHash;
+
+    public FileWatcherTask(
+        URL configURL,
+        Collection<URL> schemaTypes,
+        Function<String, Guard> guardsByType,
+        ToIntFunction<String> supplyId,
+        IntFunction<ToIntFunction<KindConfig>> maxWorkers,
+        Tuning tuning,
+        Collection<DispatchAgent> dispatchers,
+        ErrorHandler errorHandler,
+        Consumer<String> logger,
+        EngineExtContext context,
+        EngineConfiguration config,
+        List<EngineExtSpi> extensions)
+    {
+        this.configURL = configURL;
+        this.schemaTypes = schemaTypes;
+        this.guardsByType = guardsByType;
+        this.supplyId = supplyId;
+        this.maxWorkers = maxWorkers;
+        this.tuning = tuning;
+        this.dispatchers = dispatchers;
+        this.errorHandler = errorHandler;
+        this.logger = logger;
+        this.context = context;
+        this.config = config;
+        this.extensions = extensions;
+    }
+
+    @Override
+    public boolean run()
+    {
+        try
+        {
+            Path configPath = Paths.get(new File(configURL.getPath()).getAbsolutePath());
+
+            WatchService watchService = FileSystems.getDefault().newWatchService();
+            configPath.getParent().register(watchService, ENTRY_MODIFY, ENTRY_CREATE, ENTRY_DELETE);
+
+            Path configFileName = configPath.getFileName();
+            configHash = getConfigHash();
+            while (true)
+            {
+                if (rootNamespace != null)
+                {
+                    try
+                    {
+                        final WatchKey key = watchService.take();
+
+                        for (WatchEvent<?> event : key.pollEvents())
+                        {
+                            final Path changed = (Path) event.context();
+                            if (changed.equals(configFileName))
+                            {
+                                byte[] newConfigHash = getConfigHash();
+                                if (!Arrays.equals(configHash, newConfigHash))
+                                {
+                                    configHash = newConfigHash;
+                                    commonPool().submit(new UnregisterTask(dispatchers, rootNamespace, context, extensions))
+                                        .get();
+                                    rootNamespace = commonPool().submit(
+                                        new RegisterTask(configURL, schemaTypes, guardsByType, supplyId, maxWorkers, tuning,
+                                            dispatchers, errorHandler, logger, context, config, extensions)
+                                    ).get();
+                                }
+                            }
+                        }
+                        key.reset();
+                    }
+                    catch (InterruptedException ex)
```

Comment:
Just a note here: we discussed that maybe catching a ClosedWatchServiceException here would be enogh, so we don't need to store the current thread and do interrupt. However, we only get ClosedWatchServiceException, when we call take or poll on a closed WatchService. In our case, we might be already at the staet of unregistering/registering, when we got the close call, which closes the WatchService. In that case, we won't stop the detach. Tested it, and got the same NPE, when trying to close an already closed stream.